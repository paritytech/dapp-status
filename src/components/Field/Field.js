// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Form from 'semantic-ui-react/dist/commonjs/collections/Form';
import Label from 'semantic-ui-react/dist/commonjs/elements/Label';
import Popup from 'semantic-ui-react/dist/commonjs/modules/Popup';
import SemanticInput from 'semantic-ui-react/dist/commonjs/elements/Input';
import { FormattedMessage, injectIntl, intlShape } from 'react-intl';

export class Field extends Component {
  static propTypes = {
    action: PropTypes.object,
    intl: intlShape,
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    onSubmit: PropTypes.func,
    readOnly: PropTypes.bool,
    showCopyButton: PropTypes.bool, // Will be overridden if action is set
    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    width: PropTypes.number
  };

  state = {
    error: null,
    isCopied: false,
    isLoading: false,
    isSaved: false,
    value: null
  };

  componentWillReceiveProps({ value }) {
    // Reset field when new value comes
    if (this.props.value !== value && value) {
      this.setState({ value });
    }
  }

  handleChange = (_, { value }) => this.setState({ value, error: null });

  handleClearError = () => this.setState({ error: null });

  handleCopy = () => {
    // https://stackoverflow.com/questions/39501289/in-reactjs-how-to-copy-text-to-clipboard
    var tmp = document.createElement('textarea');
    tmp.innerText = this.props.value;
    document.body.appendChild(tmp);
    tmp.select();
    document.execCommand('copy');
    tmp.remove();
    this.setState({ isCopied: true });
    setTimeout(() => this.setState({ isCopied: false }), 2000);
  };

  handleKeyPress = event => {
    // Submit on enter
    if (event.keyCode === 13 || event.charCode === 13) {
      this.handleSubmit();
    }
  };

  handleSubmit = () => {
    if (!this.state.value || this.state.value === this.props.value) {
      return;
    }

    this.setState({ isLoading: true });
    this.props
      .onSubmit(this.state.value)
      .then(() => this.setState({ isLoading: false, isSaved: true }))
      .then(() => new Promise(resolve => setTimeout(resolve, 2000)))
      .then(() => this.setState({ isSaved: false }))
      .catch(error => this.setState({ isLoading: false, error }));
  };

  /**
   * Render an input which submits through an action button
   */
  renderActionInput = () => {
    const {
      action,
      intl: { formatMessage },
      label,
      readOnly,
      showCopyButton,
      value,
      ...rest
    } = this.props;

    // Customize the action of the input depending on state
    let inputAction;
    if (this.state.error) {
      inputAction = {
        ...action,
        content: formatMessage(messages.clearError),
        icon: 'remove',
        onClick: this.handleClearError,
        size: 'tiny'
      };
    } else if (this.state.isSaved) {
      inputAction = {
        ...action,
        content: formatMessage(messages.success),
        disabled: true,
        icon: 'check',
        positive: true,
        size: 'tiny'
      };
    } else {
      inputAction = {
        ...action,
        disabled: !this.state.value || this.state.isLoading,
        loading: this.state.isLoading,
        onClick: this.handleSubmit,
        primary: true,
        size: 'tiny'
      };
    }

    return (
      <SemanticInput
        action={inputAction}
        onChange={this.handleChange}
        readOnly={readOnly || this.state.isLoading || this.state.isSaved} // Don't allow changing when loading
        value={this.state.value} // Controlled component
        {...rest}
      />
    );
  };

  /**
   * Render an input with an icon, which submits through blur or keypress
   */
  renderIconInput = () => {
    const {
      action,
      label,
      readOnly,
      showCopyButton,
      value,
      ...rest
    } = this.props;
    return (
      <SemanticInput
        action={
          showCopyButton && {
            icon: 'copy',
            onClick: this.handleCopy
          }
        }
        loading={this.state.isLoading}
        icon={this.state.isSaved ? { name: 'check', color: 'green' } : true}
        onKeyPress={this.handleKeyPress}
        onBlur={this.handleSubmit}
        onChange={this.handleChange}
        readOnly={readOnly || this.state.isLoading} // Don't allow changing when loading
        value={this.state.value} // Controlled component
        {...rest}
      />
    );
  };

  renderInput = () =>
    this.props.action ? this.renderActionInput() : this.renderIconInput();

  render() {
    const { label, showCopyButton, width } = this.props;
    return (
      <Form.Field width={width}>
        <label>{label}</label>
        {showCopyButton ? (
          <Popup
            content={
              <FormattedMessage
                id="dapp.status.field.copied"
                defaultMessage="Copied"
              />
            }
            inverted
            open={this.state.isCopied}
            position="top center"
            size="mini"
            trigger={this.renderInput()}
          />
        ) : (
          this.renderInput()
        )}
        {this.state.error && (
          <Label basic color="red" pointing>
            {this.state.error.toString()}
          </Label>
        )}
      </Form.Field>
    );
  }
}

export default injectIntl(Field);

const messages = {
  clearError: {
    id: 'dapp.status.field.clearError',
    defaultMessage: 'Clear Error'
  },
  success: {
    id: 'dapp.status.field.success',
    defaultMessage: 'Success'
  }
};
