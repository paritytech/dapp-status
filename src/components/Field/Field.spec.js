// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { shallowToJson } from 'enzyme-to-json';
import Popup from 'semantic-ui-react/dist/commonjs/modules/Popup';
import SemanticInput from 'semantic-ui-react/dist/commonjs/elements/Input';

import { shallowWithIntl } from '../../setupTests';
import { Field } from './Field';

const props = { onSubmit: () => Promise.resolve(), value: 'Foo' };

describe('when we have a normal input', () => {
  test('should render correctly', () => {
    const component = shallowWithIntl(<Field />);

    expect(shallowToJson(component)).toMatchSnapshot();
  });

  test('should render correctly with copy button', () => {
    const component = shallowWithIntl(<Field {...props} showCopyButton />);

    expect(shallowToJson(component)).toMatchSnapshot();
  });

  test('should handle copy', () => {
    global.document.execCommand = jest.fn();
    const component = shallowWithIntl(<Field {...props} showCopyButton />);
    const input = component.find(Popup).props().trigger;
    input.props.action.onClick();

    expect.assertions(3);
    expect(component.state('isCopied')).toBe(true);
    expect(global.document.execCommand).toHaveBeenCalledWith('copy');
    return new Promise(resolve => setTimeout(resolve, 2100)).then(() => {
      expect(component.state('isCopied')).toBe(false);
    });
  });

  test('should clear error when state value change', () => {
    const component = shallowWithIntl(<Field {...props} />);

    component.setState({ error: 'Some Error' });
    expect(component.state('error')).toBe('Some Error');

    const input = component.find(SemanticInput);
    input.props().onChange(null, { value: 'Bar' });
    expect(component.state('error')).toBe(null);
  });

  test('should not call onSubmit on blur if state value is not set', () => {
    const onSubmit = jest.fn(() => Promise.resolve());
    const component = shallowWithIntl(<Field {...props} onSubmit={onSubmit} />);
    const input = component.find(SemanticInput);
    input.props().onBlur();

    expect(onSubmit).not.toHaveBeenCalled();
  });

  test('should not call onSubmit on blur if state value is equal to initial value', () => {
    const onSubmit = jest.fn(() => Promise.resolve());
    const component = shallowWithIntl(<Field {...props} onSubmit={onSubmit} />);
    const input = component.find(SemanticInput);
    input.props().onChange(null, { value: 'Foo' });
    input.props().onBlur();

    expect(onSubmit).not.toHaveBeenCalled();
  });

  test('should call onSubmit on blur if state value is set', () => {
    const onSubmit = jest.fn(() => Promise.resolve());
    const component = shallowWithIntl(<Field {...props} onSubmit={onSubmit} />);
    const input = component.find(SemanticInput);
    input.props().onChange(null, { value: 'Bar' });
    input.props().onBlur();

    expect(onSubmit).toHaveBeenCalledWith('Bar');
  });

  test('should call onSubmit on enter key press', () => {
    const onSubmit = jest.fn(() => Promise.resolve());
    const component = shallowWithIntl(<Field {...props} onSubmit={onSubmit} />);
    const input = component.find(SemanticInput);
    input.props().onChange(null, { value: 'Bar' });
    input.props().onKeyPress({ keyCode: 13 });

    expect(onSubmit).toHaveBeenCalledWith('Bar');
  });

  test('should not call onSubmit when keypress with another key', () => {
    const onSubmit = jest.fn(() => Promise.resolve());
    const component = shallowWithIntl(<Field {...props} onSubmit={onSubmit} />);
    const input = component.find(SemanticInput);
    input.props().onChange(null, { value: 'Bar' });
    input.props().onKeyPress({ charCode: 14 });

    expect(onSubmit).not.toHaveBeenCalled();
  });
});

describe('when we have an action input', () => {
  const action = { content: 'Foo' };

  test('should render correctly', () => {
    const component = shallowWithIntl(<Field {...props} action={action} />);

    expect(shallowToJson(component)).toMatchSnapshot();
  });

  test('should call onSubmit on action button click', () => {
    const onSubmit = jest.fn(() => Promise.resolve());
    const component = shallowWithIntl(
      <Field {...props} action={action} onSubmit={onSubmit} />
    );
    const input = component.find(SemanticInput);
    input.props().onChange(null, { value: 'Bar' });
    input.props().action.onClick();

    expect(onSubmit).toHaveBeenCalledWith('Bar');
  });

  test('should clear error on button click', () => {
    const component = shallowWithIntl(<Field {...props} action={action} />);

    component.setState({ error: 'Some Error' });
    expect(component.state('error')).toBe('Some Error');

    const input = component.find(SemanticInput);
    input.props().action.onClick();
    expect(component.state('error')).toBe(null);
  });
});

test('should change state value when changing input', () => {
  const component = shallowWithIntl(<Field {...props} />);
  const input = component.find(SemanticInput);

  input.props().onChange(null, { value: 'Bar' });
  expect(component.state('value')).toBe('Bar');
});

test('should reinitialize state value when prop value changes', () => {
  const component = shallowWithIntl(<Field {...props} />);
  component.setProps({ value: 'Bar' });

  expect(component.state('value')).toBe('Bar');
});

test('should not change state value if another prop is modified', () => {
  const component = shallowWithIntl(<Field {...props} />);
  component.setProps({ width: 2 });

  expect(component.state('value')).toBe(null);
});

test('should handle isLoading and isSaved states after submitting', () => {
  const component = shallowWithIntl(<Field {...props} />);
  const input = component.find(SemanticInput);
  input.props().onChange(null, { value: 'Bar' });
  input.props().onBlur();

  expect.assertions(4);
  expect(component.state('isLoading')).toBe(true);
  return props
    .onSubmit()
    .then(() => {
      expect(component.state('isLoading')).toBe(false);
      expect(component.state('isSaved')).toBe(true);
      return new Promise(resolve => setTimeout(resolve, 2100));
    })
    .then(() => {
      expect(component.state('isSaved')).toBe(false);
    });
});

test('should handle isLoading and error if submitting yields error', () => {
  const onSubmit = () => Promise.reject('Some Error');
  const component = shallowWithIntl(<Field {...props} onSubmit={onSubmit} />);
  const input = component.find(SemanticInput);
  input.props().onChange(null, { value: 'Bar' });
  input.props().onBlur();

  expect.assertions(3);
  expect(component.state('isLoading')).toBe(true);
  return onSubmit()
    .catch(() => new Promise(resolve => setTimeout(resolve, 100)))
    .then(() => {
      expect(component.state('isLoading')).toBe(false);
      expect(component.state('error')).toBe('Some Error');
    });
});
