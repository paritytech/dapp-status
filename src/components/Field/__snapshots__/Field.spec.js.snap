// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when we have a normal input should render correctly 1`] = `
<FormField>
  <label />
  <Input
    icon={true}
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {},
        "now": [Function],
        "textComponent": "span",
      }
    }
    loading={false}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyPress={[Function]}
    readOnly={false}
    type="text"
    value={null}
  />
</FormField>
`;

exports[`when we have a normal input should render correctly with copy button 1`] = `
<FormField>
  <label />
  <Popup
    content={
      <FormattedMessage
        defaultMessage="Copied"
        id="dapp.status.field.copied"
        values={Object {}}
      />
    }
    inverted={true}
    on="hover"
    open={false}
    position="top center"
    size="mini"
    trigger={
      <Input
        action={
          Object {
            "icon": "copy",
            "onClick": [Function],
          }
        }
        icon={true}
        intl={
          Object {
            "defaultFormats": Object {},
            "defaultLocale": "en",
            "formatDate": [Function],
            "formatHTMLMessage": [Function],
            "formatMessage": [Function],
            "formatNumber": [Function],
            "formatPlural": [Function],
            "formatRelative": [Function],
            "formatTime": [Function],
            "formats": Object {},
            "formatters": Object {
              "getDateTimeFormat": [Function],
              "getMessageFormat": [Function],
              "getNumberFormat": [Function],
              "getPluralFormat": [Function],
              "getRelativeFormat": [Function],
            },
            "locale": "en",
            "messages": Object {},
            "now": [Function],
            "textComponent": "span",
          }
        }
        loading={false}
        onBlur={[Function]}
        onChange={[Function]}
        onKeyPress={[Function]}
        onSubmit={[Function]}
        readOnly={false}
        type="text"
        value={null}
      />
    }
  />
</FormField>
`;

exports[`when we have an action input should render correctly 1`] = `
<FormField>
  <label />
  <Input
    action={
      Object {
        "content": "Foo",
        "disabled": true,
        "loading": false,
        "onClick": [Function],
        "primary": true,
        "size": "tiny",
      }
    }
    onChange={[Function]}
    onSubmit={[Function]}
    readOnly={false}
    type="text"
    value={null}
  />
</FormField>
`;
